BDArmory/Control/BDModulePilotAI.cs

```diff
@@ -2329,7 +2334,10 @@ void FlyToTargetVessel(FlightCtrlState s, Vessel v)
                         magnifier += ((magnifier - 1f) * Mathf.Sin(Time.time * 0.75f));
                         if (BDArmorySettings.DEBUG_TELEMETRY || BDArmorySettings.DEBUG_AI) debugString.AppendLine($"targetAngVel: {targetAngVel:F4}, magnifier: {magnifier:F2}");
                         target -= magnifier * leadOffset; // The effect of this is to exagerate the lead if the angular velocity is > 1
-                        angleToTarget = Vector3.Angle(vesselTransform.up, target - vesselTransform.position);
+                        //need to subtract the angle of the gun if offset/angled from the target est. position
+                        //target -= leadOffset; //correctly account for offset guns/schrage Musik
+                        target = Quaternion.FromToRotation(weapon.fireTransforms[0].forward, vesselTransform.up) * (target - vesselTransform.position) + vesselTransform.position;
+                        angleToTarget = Vector3.Angle(weapon.fireTransforms[0].transform.position, target - vesselTransform.position);
```

Comment:
This should be using `weapon.fireTransforms[0].forward` not `weapon.fireTransforms[0].transform.position`.

Testing this, it seems to work ok for one of my more agile craft. It manages to aim ok with pitch offsets somewhere between -2° — 30° and yaw offsets between ±10°.

Something else I noticed is that the lines above this involving the `magnifier` only serve to throw off the aim. I guess that's from when aiming was very approximate and a bit of wiggle was a good thing. Now, it only serves to make sniper shots miss half the time. I'd suggest we remove it (which should please the FJRT guys using the EML), i.e.,
```diff
diff --git a/BDArmory/Control/BDModulePilotAI.cs b/BDArmory/Control/BDModulePilotAI.cs
index 6c5824f6e..494855c63 100644
--- a/BDArmory/Control/BDModulePilotAI.cs
+++ b/BDArmory/Control/BDModulePilotAI.cs
@@ -2328,16 +2328,9 @@ void FlyToTargetVessel(FlightCtrlState s, Vessel v)
                     if (weapon != null)
                     {
                         Vector3 leadOffset = weapon.GetLeadOffset();
-
-                        float targetAngVel = Vector3.Angle(v.transform.position - vessel.transform.position, v.transform.position + (vessel.Velocity()) - vessel.transform.position);
-                        float magnifier = Mathf.Clamp(targetAngVel, 1f, 2f);
-                        magnifier += ((magnifier - 1f) * Mathf.Sin(Time.time * 0.75f));
-                        if (BDArmorySettings.DEBUG_TELEMETRY || BDArmorySettings.DEBUG_AI) debugString.AppendLine($"targetAngVel: {targetAngVel:F4}, magnifier: {magnifier:F2}");
-                        target -= magnifier * leadOffset; // The effect of this is to exagerate the lead if the angular velocity is > 1
-                        //need to subtract the angle of the gun if offset/angled from the target est. position
-                        //target -= leadOffset; //correctly account for offset guns/schrage Musik
-                        target = Quaternion.FromToRotation(weapon.fireTransforms[0].forward, vesselTransform.up) * (target - vesselTransform.position) + vesselTransform.position;
-                        angleToTarget = Vector3.Angle(weapon.fireTransforms[0].transform.position, target - vesselTransform.position);
+                        target -= leadOffset;
+                        target = Quaternion.FromToRotation(weapon.fireTransforms[0].forward, vesselTransform.up) * (target - vesselTransform.position) + vesselTransform.position; //correctly account for offset guns/schrage Musik
+                        angleToTarget = Vector3.Angle(weapon.fireTransforms[0].forward, target - vesselTransform.position);
                         if (distanceToTarget < weaponManager.gunRange && angleToTarget < 20) // FIXME This ought to be changed to a dynamic angle like the firing angle.
                         {
                             steerMode = SteerModes.Aiming; //steer to aim
```

In the last comment you also mentioned wanting to apply a correction for weapons that are offset by a translation too. For that case, `target` ought to also be offset by the opposite of the weapon's offset from `vesselTransform.up` prior to the rotation.

Something else I thought of is that for weapons that are symmetric, the angles and translational offsets ought to be averaged over the symmetric weapons (or maybe the entire weapon group?), e.g., if you put converging weapons on the wingtips, you don't want to be aiming with just one of them, but rather aiming down the centre to the point that they converge on.

Reply:
Doing an average of symmetric weapons would work, as those should all be pointing in the same direction.
Weapongroup is a tricky one, since you could potentially be dealing with several weapons with different ballisitic performance and lead offset requirements, and an average of them means none of them line up with a target.
Might make sense to add a 'Aim From Here' toggle to weapons to set which one is used for aiming (and if none are selected, then do an average of all vectors?)?